File: src/Unit.java
import java.util.Set;
public class Unit {
    private String name;
    private String code;
    private int credits;
    private String description;
    private RequirementRule corequisite;  //meantime or before
    private RequirementRule prerequisite; //before
    private RequirementRule prohibition;  //cannot take if completed
    private Set<TeachingPeriod> teachingPeriod;
    

    public Unit(String name, String code, int credits, String description, 
        RequirementRule corequisite, RequirementRule prerequisite, RequirementRule prohibition,
        Set<TeachingPeriod> teachingPeriod) 
    {
        this.name = name;
        this.code = code;
        this.credits = credits;
        this.description = description;
        this.corequisite = corequisite;
        this.prerequisite = prerequisite;
        this.prohibition = prohibition;
        this.teachingPeriod = teachingPeriod;
    }


    //getters
    public String getName() {
        return name;
    }

    public String getCode() {
        return code;
    }

    public int getCredits() {
        return credits;
    }

    public String getDescription() {
        return description;
    }

    public RequirementRule getCorequisite() {
        return corequisite;
    }

    public  RequirementRule getPrerequisite() {
        return prerequisite;
    }

    public RequirementRule getProhibition() {
        return prohibition;
    }

    public Set<TeachingPeriod> getTeachingPeriod() {
        return teachingPeriod;
    }



    //setters
    public void setName(String name) {
        this.name = name;
    }

    public void setCode(String code){
        this.code = code;
    }

    public void setCredits(int credits) {
        this.credits = credits;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public void setTeachingPeriod(Set<TeachingPeriod> teachingPeriod) {
        this.teachingPeriod = teachingPeriod;
    }


}

File: src/unit.cpp
#include "unit.hpp"

#include <utility>

namespace course_planner {

Unit::Unit(std::string name, std::string code, int credits,
           std::string description, RequirementRulePtr corequisite,
           RequirementRulePtr prerequisite, RequirementRulePtr prohibition,
           std::set<TeachingPeriod> teaching_periods)
    : name_(std::move(name)),
      code_(std::move(code)),
      credits_(credits),
      description_(std::move(description)),
      corequisite_(std::move(corequisite)),
      prerequisite_(std::move(prerequisite)),
      prohibition_(std::move(prohibition)),
      teaching_periods_(std::move(teaching_periods)) {}

const std::string& Unit::name() const {
    return name_;
}

const std::string& Unit::code() const {
    return code_;
}

int Unit::credits() const {
    return credits_;
}

const std::string& Unit::description() const {
    return description_;
}

const RequirementRulePtr& Unit::corequisite() const {
    return corequisite_;
}

const RequirementRulePtr& Unit::prerequisite() const {
    return prerequisite_;
}

const RequirementRulePtr& Unit::prohibition() const {
    return prohibition_;
}

const std::set<TeachingPeriod>& Unit::teaching_periods() const {
    return teaching_periods_;
}

void Unit::set_name(std::string name) {
    name_ = std::move(name);
}

void Unit::set_code(std::string code) {
    code_ = std::move(code);
}

void Unit::set_credits(int credits) {
    credits_ = credits;
}

void Unit::set_description(std::string description) {
    description_ = std::move(description);
}

void Unit::set_teaching_periods(std::set<TeachingPeriod> teaching_periods) {
    teaching_periods_ = std::move(teaching_periods);
}

}  // namespace course_planner
